hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/hdds/upgrade/TestHDDSUpgrade.java

```diff
@@ -305,4 +405,655 @@ public void testFinalizationFromInitialVersionToLatestVersion()
     // Verify that new pipeline can be created with upgraded datanodes.
     testPostUpgradePipelineCreation();
   }
+
+  /*
+   * All the subsequent tests here are failure cases. Some of the tests below
+   * could simultaneously fail one or more nodes at specific execution points
+   * and in different thread contexts.
+   * Upgrade path key execution points are defined in
+   * UpgradeFinalizer:UpgradeTestInjectionPoints.
+   */
+
+  /*
+   * Helper function to inject SCM failure and a SCM restart at a given
+   * execution point during SCM-Upgrade.
+   *
+   * Injects Failure in  : SCM
+   * Executing-Thread-Context : SCM-Upgrade
+   */
+  private Boolean injectSCMFailureDuringSCMUpgrade()
+      throws InterruptedException, TimeoutException, AuthenticationException,
+      IOException {
+    // For some tests this could get called in a different thread context.
+    // We need to guard concurrent updates to the cluster.
+    synchronized(cluster) {
+      cluster.restartStorageContainerManager(true);
+      loadSCMState();
+    }
+    // The ongoing current SCM Upgrade is getting aborted at this point. We
+    // need to schedule a new SCM Upgrade on a different thread context.
+    Thread t = new Thread(new Runnable() {
+      @Override
+      public void run() {
+        try {
+          loadSCMState();
+          scm.finalizeUpgrade("xyz");
```

Comment:
Thats because, this finalization will fail and we need another thread to simulate another attempt at failed finalization.

Reply:
@prashantpogde Maybe I am missing something here, but the method 'testFinalizationWithFailuerInjectionHelper' is also invoked, and within that we do a _scm.finalizeUpgrade("xyz")_. Why do we need to finalize twice?

Reply:
I see, removed this.

Reply:
Spoke too soon. This can not be removed. 

- Test cases  Where it is not a must, it validates that the multiple invocations of finalizations are handled gracefully.
- It is a must for TestSCMFailureAfterCompleteFinalizaion. This excercises SCM failure just before postFinalization. This case case not be be caught in "*Helper"Function and is race-prone. The only guaranteed way to address it is in this thread, at this execution point.